### PR TITLE
Functions  to fund smart wallet or P2P address with either the native token or a specific ERC20

### PR DESCRIPTION
**Pull request by riccardopillar**

### CODE REVIEW

File: .gitignore [hunk 1]
```diff
@@ -14,6 +14,7 @@
 !.vscode/tasks.json
 !.vscode/launch.json
 !.vscode/extensions.json
+workspace.code-workspace
```
Comment (zincoshine) [inline .gitignore:17]: Pls don't push workspace to the repo.

Comment (zincoshine) [reply .gitignore:17]: Ah wait this is to .gitignore, then I guess that's alright.

File: src/sdk/sdk.ts [hunk 2]
```diff
@@ -2028,6 +2028,91 @@ export class Sdk {
...
+      "outputs": [],
+      "payable": false
+    }
+  ];
+
+  async topUpToken(
+    value: string,
+    contractAddress: string
+  ): Promise<void> {
+    const account = this.state.p2pPaymentDepositAddress;
```
Comment (zincoshine) [inline src/sdk/sdk.ts:2086]: This must be the accountAddress

File: src/sdk/sdk.ts [hunk 3]
```diff
@@ -2028,6 +2028,91 @@ export class Sdk {
...
+    await this.transferTokens(account, value, contractAddress)
+  }
+
+  async topUpTokenP2P(
+    value: string,
+    contractAddress: string
+  ): Promise<void> {
+    const account = this.state.accountAddress;
```
Comment (zincoshine) [inline src/sdk/sdk.ts:2094]: This has to be the P2P address.

File: src/sdk/sdk.ts [hunk 4]
```diff
@@ -2028,6 +2028,91 @@ export class Sdk {
...
+    const provider = this.services.walletService.walletProvider as any;
+    const numberOfTokens = utils.parseUnits(value, 18)
+    if (provider) {
+      const contract = new EthersContract(
+        contractAddress,
+        this.contractAbiFragment,
+        provider
+      )
+      contract.transfer(account, numberOfTokens).then(() => {
+        // console.dir(transferResult)
```
Comment (zincoshine) [inline src/sdk/sdk.ts:2108]: Please remove unused code. Should we replace it with a proper return message?

File: src/sdk/sdk.ts [hunk 5]
```diff
@@ -2028,6 +2028,91 @@ export class Sdk {
...
+    value: string,
+    contractAddress: string
+  ): Promise<void> {
+    const account = this.state.accountAddress;
+    await this.transferTokens(account, value, contractAddress)
+  }
+
+  private async transferTokens(account: string, value: string, contractAddress): Promise<void> {
+    const provider = this.services.walletService.walletProvider as any;
+    const numberOfTokens = utils.parseUnits(value, 18)
```
Comment (rafaelcidade) [inline src/sdk/sdk.ts:2100]: Token decimals should not be hardcoded

File: src/sdk/sdk.ts [hunk 6]
```diff
@@ -2028,6 +2028,91 @@ export class Sdk {
     return this.services.faucetService.topUpPaymentDepositAccount();
   }
 
+
+  async topUp(
+    value: string,
+  ): Promise<void> {
+    const wallet: Partial<Wallet> = this.services.walletService.walletProvider;
```
Comment (zincoshine) [inline src/sdk/sdk.ts:2035]: How are we handling null here? What if walletProvider is null? Please add appropriate try/catch

File: src/sdk/sdk.ts (hunk 3 above)
Comment (rafaelcidade) [inline src/sdk/sdk.ts:2094]: `this.state.p2pPaymentDepositAddress`

File: src/sdk/sdk.ts (hunk 2 above)
Comment (rafaelcidade) [inline src/sdk/sdk.ts:2086]: `this.state.accountAddress`

File: workspace.code-workspace [hunk 7]
```diff
@@ -0,0 +1,20 @@
+{
```
Comment (zincoshine) [inline workspace.code-workspace:1]: This file has to be deleted from the PR

**Review by zincoshine — CHANGES_REQUESTED** [review]

File: src/sdk/account/account.service.ts [hunk 8]
```diff
@@ -485,6 +485,10 @@ export class AccountService extends Service {
     return result;
   }
 
+  isAccountTypeContract():boolean {
```
Comment (zincoshine) [inline src/sdk/account/account.service.ts:488]: I would rename the function as isContractAccount()

**Review by zincoshine — CHANGES_REQUESTED** [review]

File: src/sdk/sdk.ts [hunk 9]
```diff
@@ -2028,6 +2028,103 @@ export class Sdk {
...
+        {
+          "type": "uint256",
+          "name": "_tokens"
+        }
+      ],
+      "constant": false,
+      "outputs": [],
+      "payable": false
+    }
+  ];
```
Comment (SaltyCucumber) [inline src/sdk/sdk.ts:2086]: maybe getContractAbi can help, look it up in the sdk code

File: src/sdk/sdk.ts [hunk 10]
```diff
@@ -2028,6 +2028,103 @@ export class Sdk {
...
+  ): Promise<void> {
+    if (!this.services.accountService.isContractAccount()) await this.computeContractAccount({
+      sync: false,
+    });
+
+    const account = this.state.p2pPaymentDepositAddress;
+    await this.transferTokens(account, value, contractAddress)
+  }
+
+  private async transferTokens(account: string, value: string, contractAddress: string, decimals = 18): Promise<void> {
```
[the remark this answers is not in the document]
Comment (riccardopillar) [reply src/sdk/sdk.ts:2085]: I guess is more safe remove the default value

File: src/sdk/sdk.ts [hunk 11]
```diff
@@ -2028,6 +2028,103 @@ export class Sdk {
     return this.services.faucetService.topUpPaymentDepositAccount();
   }
 
+
+  async topUp(
+    value: string,
+  ): Promise<void> {
+    if (this.services.accountService.isContractAccount()) await this.computeContractAccount({
```
[the remark this answers is not in the document]
Comment (riccardopillar) [reply src/sdk/sdk.ts]: actually the condition should be if not, was an error from my side

**Review by rafaelcidade — APPROVED** [review]

**Review by zincoshine — APPROVED** [review]

**Review by SaltyCucumber — APPROVED** [review]

File: .gitignore (hunk 1 above)
[the remark this answers is not in the document]
Comment (riccardopillar) [reply .gitignore:17]: I have added it for avoid to commit the file

File: examples/src/03-contract-account.ts [hunk 12]
```diff
@@ -4,7 +4,7 @@ import { logger, topUpAccount, getBalance, randomAddress } from './common';
 
 async function main(): Promise<void> {
   const wallet = Wallet.createRandom();
-
+
```
[the remark this answers is not in the document]
Comment (riccardopillar) [reply examples/src/03-contract-account.ts:7]: I have run it, it seems it doesn't

**Pull request merged**